sdk/couchbase-core/src/queryx/query.rs

```diff
@@ -138,4 +147,538 @@ impl<C: Client> Query<C> {
+                    temp_where, encoded_bucket
+                )
+            } else {
+                temp_where
+            }
+        } else if opts.collection_name.is_none() && opts.scope_name.is_some() {
+            let scope_name = normalise_default_name(opts.scope_name.as_ref().unwrap());
+
+            let encoded_bucket = encode_value(&opts.bucket_name)?;
+            let encoded_scope = encode_value(&scope_name)?;
+
+            format!(
+                "bucket_id={} AND scope_id={}",
+                encoded_bucket, encoded_scope
+            )
+        } else {
+            return Err(Error::new_generic_error(
+                "Invalid combination of bucket, scope and collection".to_string(),
+                &self.endpoint,
+                "",
+                "",
+            ));
+        };
+
+        where_clause = format!("({}) AND `using`=\"gsi\"", where_clause);
+        let qs = format!(
+            "SELECT `idx`.* FROM system:indexes AS idx WHERE {} ORDER BY is_primary DESC, name ASC",
+            where_clause
+        );
+
+        let opts = QueryOptions::new()
+            .statement(qs)
+            .on_behalf_of(opts.on_behalf_of.clone());
+        let mut res = self.query(&opts).await?;
+
+        let mut indexes = vec![];
+
+        while let Some(row) = res.next().await {
+            let bytes = row?;
+            let index: Index = serde_json::from_slice(&bytes)
+                .map_err(|e| Error::new_generic_error(e.to_string(), "", "", ""))?;
+
+            indexes.push(index);
+        }
+
+        Ok(indexes)
+    }
+
+    pub async fn create_primary_index(
+        &self,
+        opts: &CreatePrimaryIndexOptions,
+    ) -> error::Result<()> {
+        let mut qs = String::from("CREATE PRIMARY INDEX");
+        if opts.index_name.is_some() {
+            qs.push_str(&format!(
+                " {}",
+                encode_identifier(opts.index_name.as_ref().unwrap())
+            ));
+        }
+
+        qs.push_str(&format!(
+            " ON {}",
+            build_keyspace(
+                opts.bucket_name.as_ref(),
+                &opts.scope_name.as_deref(),
+                &opts.collection_name.as_deref()
+            )
+        ));
+
+        let mut with: HashMap<&str, Value> = HashMap::new();
+
+        if let Some(deferred) = opts.deferred {
+            with.insert("defer_build", Value::Bool(deferred));
+        }
+
+        if let Some(num_replicas) = opts.num_replicas {
+            with.insert("num_replica", Value::Number(num_replicas.into()));
+        }
+
+        if !with.is_empty() {
+            let with = encode_value(&with)?;
+            qs.push_str(&format!(" WITH {}", with));
+        }
+
+        let query_opts = QueryOptions::new()
+            .statement(qs)
+            .on_behalf_of(opts.on_behalf_of.clone());
+
+        let mut res = self.query(&query_opts).await;
+
+        // This is a bit mad...
+        match res {
```

Comment:
We should check with PM which server versions we need to support, statements with IF NOT EXISTS are supported as of 7.1 and 7.0 is already end of maintenance and EOL in July - so is 7.1 actually.